workflow/controller/operator.go

```diff
@@ -2770,6 +2770,25 @@ func (woc *wfOperationCtx) markNodeWaitingForLock(nodeName string, lockName stri
 	return node, nil
 }
 
+func (woc *wfOperationCtx) findLeafNodeWithType(boundaryID string, nodeType wfv1.NodeType) *wfv1.NodeStatus {
+	var leafNode *wfv1.NodeStatus
+	var dfs func(nodeID string)
+	dfs = func(nodeID string) {
+		node, err := woc.wf.Status.Nodes.Get(nodeID)
+		if err != nil {
+			return
+		}
+		if node.Type == nodeType {
+			leafNode = node
+		}
+		for _, childID := range node.Children {
+			dfs(childID)
+		}
+	}
+	dfs(boundaryID)
```

Comment:
Can this recursion have some restrictions?

Reply:
I don't think we need to add any additional restrictions here, as the number of recursive calls is limited by the number of nodes, which is already limited by [Maximum Recursion Depth](https://argo-workflows.readthedocs.io/en/latest/scaling/#maximum-recursion-depth).